pr_agent/git_providers/bitbucket_server_provider.py

```diff
@@ -134,12 +140,50 @@ def get_files(self):
         diffstat = [change["path"]['toString'] for change in changes]
         return diffstat
 
+    #gets the best common ancestor: https://git-scm.com/docs/git-merge-base
+    @staticmethod
+    def get_best_common_ancestor(source_commits_list, destination_commits_list, guaranteed_common_ancestor) -> str:
+        destination_commit_hashes = {commit['id'] for commit in destination_commits_list} | {guaranteed_common_ancestor}
+
+        for commit in source_commits_list:
+            for parent_commit in commit['parents']:
+                if parent_commit['id'] in destination_commit_hashes:
+                    return parent_commit['id']
+
+        return guaranteed_common_ancestor
+
     def get_diff_files(self) -> list[FilePatchInfo]:
         if self.diff_files:
             return self.diff_files
 
-        base_sha = self.pr.toRef['latestCommit']
-        head_sha = self.pr.fromRef['latestCommit']
+        source_commits_list = list(self.bitbucket_client.get_pull_requests_commits(
+            self.workspace_slug,
+            self.repo_slug,
+            self.pr_num
+        ))
+
+        # defaults to basic diff functionality with a guaranteed common ancestor
+        base_sha, head_sha = source_commits_list[-1]['parents'][0]['id'], source_commits_list[0]['id']
+
+        # if Bitbucket api version is greater than or equal to 7.0 then use 2-way diff functionality for the base_sha
+        if self.bitbucket_api_version is not None and self.bitbucket_api_version >= LooseVersion("7.0"):
```

Comment:
it may be more clear to split this into 3 cases:
<7
<8.16
8.16+

Reply:
Unfortunately I don't think this block can be cleanly split into 3 cases like that based on the shared logic between certain cases. I do think I can cut out a call to get the source_commits_list for the >8.16 condition though so I'll at least add that